File: CSMApps/Vertebrae/v1.0.0/src/main/java/com/carrpod/vertebrae/storage/SessionStorageManager.java
package com.carrpod.vertebrae.storage;

import android.content.Context;
import android.util.Log;

import com.carrpod.vertebrae.model.KilosSession;
import com.carrpod.vertebrae.model.SessionGroup;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class SessionStorageManager {

    private static final String TAG = "SessionStorageManager";
    private static SessionStorageManager instance;
    private final Context context;
    private final Gson gson = new Gson();

    private final File baseDir;
    private final File sessionsFile;
    private final File groupsFile;

    private final Map<String, KilosSession> sessionsCache = new ConcurrentHashMap<>();
    private final Map<String, SessionGroup> groupsCache = new ConcurrentHashMap<>();

    private SessionStorageManager(Context context) {
        this.context = context.getApplicationContext();
        this.baseDir = new File(context.getExternalFilesDir(null), "Vertebrae/Sessions");
        this.sessionsFile = new File(baseDir, "sessions.json");
        this.groupsFile = new File(baseDir, "groups.json");
        baseDir.mkdirs();
        loadCache();
    }

    public static synchronized void initialize(Context context) {
        if (instance == null) {
            instance = new SessionStorageManager(context);
        }
    }

    public static SessionStorageManager getInstance() {
        if (instance == null) {
            throw new IllegalStateException("SessionStorageManager not initialized. Call initialize() first.");
        }
        return instance;
    }

    private void loadCache() {
        try {
            if (sessionsFile.exists()) {
                Type listType = new TypeToken<ArrayList<KilosSession>>() {}.getType();
                List<KilosSession> sessions = gson.fromJson(readFile(sessionsFile), listType);
                if (sessions != null) {
                    for (KilosSession session : sessions) {
                        sessionsCache.put(session.getId(), session);
                    }
                }
            }

            if (groupsFile.exists()) {
                Type listType = new TypeToken<ArrayList<SessionGroup>>() {}.getType();
                List<SessionGroup> groups = gson.fromJson(readFile(groupsFile), listType);
                if (groups != null) {
                    for (SessionGroup group : groups) {
                        groupsCache.put(group.getId(), group);
                    }
                }
            }

            // Ensure default group exists
            if (!groupsCache.containsKey("default")) {
                SessionGroup defaultGroup = new SessionGroup("default", "Default Group", 0xFF74B9FF);
                groupsCache.put("default", defaultGroup);
                saveGroups();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error loading cache", e);
        }
    }

    private String readFile(File file) {
        try {
            FileInputStream fis = new FileInputStream(file);
            byte[] data = new byte[(int) file.length()];
            fis.read(data);
            fis.close();
            return new String(data, "UTF-8");
        } catch (Exception e) {
            Log.e(TAG, "Error reading file: " + file.getName(), e);
            return "";
        }
    }

    private void writeFile(File file, String content) {
        try {
            FileOutputStream fos = new FileOutputStream(file);
            fos.write(content.getBytes("UTF-8"));
            fos.close();
        } catch (Exception e) {
            Log.e(TAG, "Error writing file: " + file.getName(), e);
        }
    }

    // Session methods
    public void saveSession(KilosSession session) {
        sessionsCache.put(session.getId(), session);
        saveSessions();
    }

    public KilosSession loadSession(String id) {
        return sessionsCache.get(id);
    }

    public List<KilosSession> loadAllSessions() {
        return new ArrayList<>(sessionsCache.values());
    }

    public List<KilosSession> loadSessionsByGroup(String groupId) {
        List<KilosSession> result = new ArrayList<>();
        for (KilosSession session : sessionsCache.values()) {
            if (groupId.equals(session.getGroupId())) {
                result.add(session);
            }
        }
        return result;
    }

    public void deleteSession(String id) {
        sessionsCache.remove(id);
        saveSessions();
    }

    private void saveSessions() {
        try {
            String json = gson.toJson(new ArrayList<>(sessionsCache.values()));
            writeFile(sessionsFile, json);
        } catch (Exception e) {
            Log.e(TAG, "Error saving sessions", e);
        }
    }

    // Group methods
    public void saveGroup(SessionGroup group) {
        groupsCache.put(group.getId(), group);
        saveGroups();
    }

    public SessionGroup loadGroup(String id) {
        return groupsCache.get(id);
    }

    public List<SessionGroup> loadAllGroups() {
        return new ArrayList<>(groupsCache.values());
    }

    public void deleteGroup(String id) {
        if (!"default".equals(id)) {
            groupsCache.remove(id);
            // Move sessions to default group
            for (KilosSession session : sessionsCache.values()) {
                if (id.equals(session.getGroupId())) {
                    KilosSession updated = session.copyWith(null, null, null, null, null, null, null, null, null, null);
                    updated.setGroupId("default");
                    sessionsCache.put(session.getId(), updated);
                }
            }
            saveGroups();
            saveSessions();
        }
    }

    private void saveGroups() {
        try {
            String json = gson.toJson(new ArrayList<>(groupsCache.values()));
            writeFile(groupsFile, json);
        } catch (Exception e) {
            Log.e(TAG, "Error saving groups", e);
        }
    }

    // File storage for inter-session transfers
    public File getSessionStorageDir(String sessionId) {
        File dir = new File(baseDir, "session_" + sessionId);
        dir.mkdirs();
        return dir;
    }

    public File getGroupStorageDir(String groupId) {
        File dir = new File(baseDir, "group_" + groupId);
        dir.mkdirs();
        return dir;
    }

    public File saveReceivedFile(String groupId, String fileName, InputStream inputStream) {
        File targetFile = new File(getGroupStorageDir(groupId), fileName);
        try {
            FileOutputStream fos = new FileOutputStream(targetFile);
            byte[] buffer = new byte[8192];
            int len;
            while ((len = inputStream.read(buffer)) > 0) {
                fos.write(buffer, 0, len);
            }
            fos.close();
        } catch (Exception e) {
            Log.e(TAG, "Error saving file", e);
        }
        return targetFile;
    }

    public File getFile(String groupId, String fileName) {
        File file = new File(getGroupStorageDir(groupId), fileName);
        return file.exists() ? file : null;
    }

    public List<File> listFiles(String groupId) {
        File dir = getGroupStorageDir(groupId);
        File[] files = dir.listFiles();
        return files != null ? new ArrayList<>(List.of(files)) : new ArrayList<>();
    }

    // Settings
    public static class Settings {
        public int heartbeatIntervalSeconds = 30;
        public int interSessionPort = 8888;
        public boolean autoReconnect = true;
        public boolean keepScreenOn = true;
        public boolean showNotifications = true;
        public String storagePath = "";
    }

    private final File settingsFile = new File(baseDir, "settings.json");

    public void saveSettings(Settings settings) {
        writeFile(settingsFile, gson.toJson(settings));
    }

    public Settings loadSettings() {
        if (settingsFile.exists()) {
            return gson.fromJson(readFile(settingsFile), Settings.class);
        }
        return new Settings();
    }
}

File: CSMApps/Vertebrae/v1.0.0/src/main/kotlin_backup/SessionStorageManager.kt
package com.carrpod.vertebrae.storage

import android.content.Context
import com.carrpod.vertebrae.model.KilosSession
import com.carrpod.vertebrae.model.SessionGroup
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.lang.reflect.Type
import java.util.concurrent.ConcurrentHashMap

class SessionStorageManager(private val context: Context) {

    private val gson = Gson()
    private val baseDir: File
    private val sessionsFile: File
    private val groupsFile: File
    private val sessionsCache = ConcurrentHashMap<String, KilosSession>()
    private val groupsCache = ConcurrentHashMap<String, SessionGroup>()

    init {
        baseDir = File(context.getExternalFilesDir(null), "Vertebrae/Sessions").apply { mkdirs() }
        sessionsFile = File(baseDir, "sessions.json")
        groupsFile = File(baseDir, "groups.json")
        loadCache()
    }

    private fun loadCache() {
        if (sessionsFile.exists()) {
            val type = object : TypeToken<MutableList<KilosSession>>() {}.type
            val list = gson.fromJson(sessionsFile.readText(), type) ?: mutableListOf()
            list.forEach { sessionsCache[it.id] = it }
        }
        if (groupsFile.exists()) {
            val type = object : TypeToken<MutableList<SessionGroup>>() {}.type
            val list = gson.fromJson(groupsFile.readText(), type) ?: mutableListOf()
            list.forEach { groupsCache[it.id] = it }
        }
        // Ensure default group exists
        if (!groupsCache.containsKey("default")) {
            val defaultGroup = SessionGroup(id = "default", name = "Default Group", color = "#74B9FF")
            groupsCache["default"] = defaultGroup
            saveGroups()
        }
    }

    suspend fun saveSession(session: KilosSession) = withContext(Dispatchers.IO) {
        sessionsCache[session.id] = session
        saveSessions()
    }

    suspend fun loadSession(id: String): KilosSession? = withContext(Dispatchers.IO) {
        sessionsCache[id]
    }

    suspend fun loadAllSessions(): List<KilosSession> = withContext(Dispatchers.IO) {
        sessionsCache.values.toList()
    }

    suspend fun loadSessionsByGroup(groupId: String): List<KilosSession> = withContext(Dispatchers.IO) {
        sessionsCache.values.filter { it.groupId == groupId }.toList()
    }

    suspend fun deleteSession(id: String) = withContext(Dispatchers.IO) {
        sessionsCache.remove(id)
        saveSessions()
    }

    private fun saveSessions() {
        sessionsFile.writeText(gson.toJson(sessionsCache.values.toList()))
    }

    suspend fun saveGroup(group: SessionGroup) = withContext(Dispatchers.IO) {
        groupsCache[group.id] = group
        saveGroups()
    }

    suspend fun loadGroup(id: String): SessionGroup? = withContext(Dispatchers.IO) {
        groupsCache[id]
    }

    suspend fun loadAllGroups(): List<SessionGroup> = withContext(Dispatchers.IO) {
        groupsCache.values.toList()
    }

    suspend fun deleteGroup(id: String) = withContext(Dispatchers.IO) {
        if (id != "default") {
            groupsCache.remove(id)
            // Move sessions to default group
            sessionsCache.values.filter { it.groupId == id }
                .forEach { sessionsCache[it.id] = it.copyWith(groupId = "default") }
            saveGroups()
            saveSessions()
        }
    }

    private fun saveGroups() {
        groupsFile.writeText(gson.toJson(groupsCache.values.toList()))
    }

    suspend fun getSessionStorageDir(sessionId: String): File = withContext(Dispatchers.IO) {
        File(baseDir, "session_$sessionId").apply { mkdirs() }
    }

    suspend fun getGroupStorageDir(groupId: String): File = withContext(Dispatchers.IO) {
        File(baseDir, "group_$groupId").apply { mkdirs() }
    }

    companion object {
        @Volatile private var INSTANCE: SessionStorageManager? = null

        fun initialize(context: Context) {
            if (INSTANCE == null) {
                synchronized(this) {
                    if (INSTANCE == null) {
                        INSTANCE = SessionStorageManager(context.applicationContext)
                    }
                }
            }
        }

        fun getInstance(): SessionStorageManager {
            return INSTANCE ?: throw IllegalStateException("SessionStorageManager not initialized")
        }
    }
}